dronegen/container_images.go

```diff
@@ -0,0 +1,942 @@
+}
+
+func NewEcrContainerRepo(accessKeyIDSecret, secretAccessKeySecret, domain string, isProduction, isImmutable, guaranteeUnique bool) *ContainerRepo {
+	nameSuffix := "staging"
+	ecrRegion := StagingEcrRegion
+	loginSubcommand := "ecr"
+	if isProduction {
+		nameSuffix = "production"
+		ecrRegion = PublicEcrRegion
+		loginSubcommand = "ecr-public"
+	}
+
+	registryOrg := ProductionRegistryOrg
+	if configureForPRTestingOnly {
+		accessKeyIDSecret = testingSecretPrefix + accessKeyIDSecret
+		secretAccessKeySecret = testingSecretPrefix + secretAccessKeySecret
+		registryOrg = testingECRRegistryOrg
+
+		if !isProduction {
+			domain = testingECRDomain
+			ecrRegion = testingECRRegion
+		}
+	}
+
+	return &ContainerRepo{
+		Name:             fmt.Sprintf("ECR - %s", nameSuffix),
+		IsProductionRepo: isProduction,
+		IsImmutable:      isImmutable,
+		Environment: map[string]value{
+			"AWS_ACCESS_KEY_ID": {
+				fromSecret: accessKeyIDSecret,
+			},
+			"AWS_SECRET_ACCESS_KEY": {
+				fromSecret: secretAccessKeySecret,
+			},
+		},
+		RegistryDomain: domain,
+		RegistryOrg:    registryOrg,
+		LoginCommands: []string{
+			"apk add --no-cache aws-cli",
+			"TIMESTAMP=$(date -d @\"$DRONE_BUILD_CREATED\" '+%Y%m%d%H%M')",
+			fmt.Sprintf("aws %s get-login-password --region=%s | docker login -u=\"AWS\" --password-stdin %s", loginSubcommand, ecrRegion, domain),
+		},
+		TagBuilder: func(tag *ImageTag) *ImageTag {
+			if guaranteeUnique {
+				tag.AppendString("$TIMESTAMP")
+			}
+
+			return tag
+		},
+	}
+}
+
+func NewQuayContainerRepo(dockerUsername, dockerPassword string) *ContainerRepo {
+	registryOrg := ProductionRegistryOrg
+	if configureForPRTestingOnly {
+		dockerUsername = testingSecretPrefix + dockerUsername
+		dockerPassword = testingSecretPrefix + dockerPassword
+		registryOrg = testingQuayRegistryOrg
+	}
+
+	return &ContainerRepo{
+		Name:             "Quay",
+		IsProductionRepo: true,
+		IsImmutable:      false,
+		Environment: map[string]value{
+			"QUAY_USERNAME": {
+				fromSecret: dockerUsername,
+			},
+			"QUAY_PASSWORD": {
+				fromSecret: dockerPassword,
+			},
+		},
+		RegistryDomain: ProductionRegistryQuay,
+		RegistryOrg:    registryOrg,
+		LoginCommands: []string{
+			fmt.Sprintf("docker login -u=\"$QUAY_USERNAME\" -p=\"$QUAY_PASSWORD\" %q", ProductionRegistryQuay),
+		},
+	}
+}
+
+func GetStagingContainerRepo(uniqueStagingTag bool) *ContainerRepo {
+	return NewEcrContainerRepo("STAGING_TELEPORT_DRONE_USER_ECR_KEY", "STAGING_TELEPORT_DRONE_USER_ECR_SECRET", StagingRegistry, false, true, uniqueStagingTag)
+}
+
+func GetProductionContainerRepos() []*ContainerRepo {
+	return []*ContainerRepo{
+		NewQuayContainerRepo("PRODUCTION_QUAYIO_DOCKER_USERNAME", "PRODUCTION_QUAYIO_DOCKER_PASSWORD"),
+		NewEcrContainerRepo("PRODUCTION_TELEPORT_DRONE_USER_ECR_KEY", "PRODUCTION_TELEPORT_DRONE_USER_ECR_SECRET", ProductionRegistry, true, false, false),
```

Comment:
ECR Production is not immutable. (Immutability on ECR - Public isn't available currently)

Reply:
Right - the `isImmutable` parameter is set to false

Reply:
@logand22 I think I see where the confusion is here. Which of the following is the expected behavior?
1. When pushing to production registries (i.e. `public.ecr.aws`) the push step should fail if an immutable tag (currently only canonical) exists.
2. When pushing to production registries (i.e. `public.ecr.aws`) the push step should log and continue if an immutable tag (currently only canonical) exists.
3. When pushing to production registries (i.e. `public.ecr.aws`) the push step should push regardless of if an immutable tag (currently only canonical) exists.

Currently I have (2) configured which is why you see lines in `.drone.yml` [like this](https://github.com/gravitational/teleport/pull/16688/files#diff-b54b39f1afced2465e1f3641db9d5bbf4f3a7fcf890996dfedd3c197bcb7f8c7R7825-R7851). I can implement any of these behaviors - just need to know which one is correct.

Reply:
Public.ecr.aws does not support tag immutability currently, therefore there is no error when attempting to push an image tag that already exists. If you want to not overwrite a canonical image when it already exists with a sort of client-side immutability then you can although I don't have that existing in the pipeline currently. 

So both 2. and 3. could work however I don't think you should fail if a tag already exists so not 1.